### DISCUSSION TITLE
Using coalesce to select vertex that may or may not be connected via an edge.

### DISCUSSION THREAD

**Discussion by musicalmathmind** [Gremlinq/ExRam.Gremlinq#828] [General]
Hello, I have the following situation.

Vertex `a` and Vertex `b` could be related via an edge called `Related`. I'm trying to build a traversal to return (`a`, `b`) if it is connected by `Related`, otherwise just return (`a`, null). We can assume `a` always exists, but it may not have a connection to `b`. My strategy was to use a `coalesce` to try and traverse from `a` to `b`. If that doesn't succeed in finding vertex `b`, I'd like to just return `Tuple<A,B>(a, null)`.

Here is what I have so far, but it has a few problems.

```
var result = await _g
                .V<A>(id)
                .As("a")
                .Coalesce<IMapGremlinQuery<(A, B)>>(
                        c1 => c1.OutE<Related>()
                            .InV<B>().As("b").Select<A, B>("a", "b"),
                        c1 => c1.Constant("").As("b").Select<A, B>("a", "b")
                ).FirstOrDefaultAsync();
```

Our team is fairly new to using Gremlin and graph databases in general, but this seems like it would work for what we're trying to accomplish. The problems we're encountering are as follows.

In the happy case (a `Related` edge exists between `a` and `b`), I do get a tuple back, however the only property that is filled out on the `B` object is it's Id. There are other properties that should be on there too, but they aren't filled out (even though they do exist as properties on the vertex). What's interesting though is if I remove the 2nd traversal in the Coalesce (the one creating the `null` constant, everything works properly for the happy case, but then the unhappy case fails by returning `(null, null)` instead of returning `(a, null)`

Any help would be greatly appreciated.

**Comment by danielcweber** [comment]
I think you're looking for something like

```
var result = await _g
    .V<A>(id)
    .Project(b => b
        .ToTuple()
        .By(__ => __)
        .By(__ => __
           .OutE<Related>()
           .OfType<B>()
           .Limit(1)
           .Fold()))
````

Syntax might be a bit off but you get the idea. The null-handling story is a bit tricky and inconsistent in Tinkerpop, so I suggest making the "b"-part of the tuple a (possibly empty) array. If it has to be a single, possibly null, element, there is something called "empty projection protection" in later versions of Gremlinq, shown [here](https://github.com/ExRam/ExRam.Gremlinq/blob/d63ece91dc6c1688e6af784004c3fe5ab3e399a9/test/ExRam.Gremlinq.Core.Tests/QueryExecutionTest.cs#L2989). You can omit the limit and fold then.

Also, I am offering workshops on Gremlin/Gremlinq. Let me know if that's something for your team.

**Reply by musicalmathmind** [reply]
Hey Daniel, thanks for responding to this - this is exactly what I was looking for. As for the workshop, would love to learn more! Do you have more information about it posted somewhere? If not you can send me an email at dthomas at felux dot com. We have a team of 3-4 engineers that would likely attend.
